src/Processors/QueryPlan/Optimizations/joinOrder.cpp

```diff
@@ -351,8 +392,115 @@ std::shared_ptr<DPJoinEntry> JoinOrderOptimizer::solveGreedy()
+    return (participating & left) && (participating & right);
+}
+
+std::shared_ptr<DPJoinEntry> JoinOrderOptimizer::solveDPsize()
 {
+    const size_t total_relations_count = query_graph.relation_stats.size();
+
+    /// Components by size (index 0 is not used that why the size is N+1)
+    std::vector<std::unordered_map<BitSet, DPJoinEntryPtr>> components(total_relations_count + 1);
+
+    /// Populate DP table for components of size=1
+    for (size_t i = 0; i < total_relations_count; ++i)
+    {
+        const auto & rel = query_graph.relation_stats[i];
+        auto entry = std::make_shared<DPJoinEntry>(i, rel.estimated_rows);
+        components[1][entry->relations] = entry;
+        dp_table[entry->relations] = entry;
+    }
+
+    /// Iteratively build components of size from 2 to N
+    for (size_t component_size = 2; component_size <= total_relations_count; ++component_size)
+    {
+        for (size_t smaller_component_size = 1; smaller_component_size <= component_size / 2; ++smaller_component_size)
+        {
+            const size_t bigger_component_size = component_size - smaller_component_size;
+
+            for (const auto & [_, right] : components[smaller_component_size])
+            {
+                checkLimits();
+
+                for (const auto & [_, left] : components[bigger_component_size])
+                {
+                    /// Do components overlap?
+                    if (left->relations & right->relations)
+                        continue;
+
+                    const auto combined_relations = left->relations | right->relations;
+
+                    auto join_kind = isValidJoinOrder(left->relations, right->relations);
+                    if (!join_kind)
+                        continue;
+
+                    /// FIXME: Restrict to Inner joins for now because isValidJoinOrder seems to not handle non-Inner joins with swapped inputs correctly
+                    if (*join_kind != JoinKind::Inner)
+                        continue;
+
+                    auto applicable_edge = getApplicableExpressions(left->relations, right->relations);
+                    /// Only leave the edges that connect left and right
+                    std::vector<JoinActionRef *> edge;
+                    for (auto & edge_it : applicable_edge)
+                    {
+                        if (connects(edge_it, left->relations, right->relations))
+                        {
+                            LOG_TEST(log, "Adding predicate connecting {} and {} : {}", left->dump(), right->dump(), edge_it->dump());
+                            edge.push_back(edge_it);
+                        }
+                        else if ((edge_it->fromLeft() || edge_it->fromRight() || edge_it->fromNone()) && component_size == 2)
+                        {
+                            /// If a predicate does not connect tables we add it at the earliest stage - when joining just 2 tables
```

Comment:
So, this guaranees that we won't apply same edge twice? In greedy we nullify applied edges, here we cannot do ot since we need them for different plan variants, right?